storage/bucket.go

```diff
@@ -264,46 +261,68 @@ func (b *BucketHandle) SignedURL(object string, opts *SignedURLOptions) (string,
 			var sa struct {
 				PrivateKey string `json:"private_key"`
 			}
-			// Don't error out if we can't marshal, fallback to GCE check.
 			err := json.Unmarshal(b.c.creds.JSON, &sa)
 			if err == nil && sa.PrivateKey != "" {
 				newopts.PrivateKey = []byte(sa.PrivateKey)
 			}
 		}
+
+		// Don't error out if we can't unmarshal the private key from the client,
+		// fallback to the default sign function for the service account.
 		if len(newopts.PrivateKey) == 0 {
 			newopts.SignBytes = b.defaultSignBytesFunc(newopts.GoogleAccessID)
 		}
 	}
 	return SignedURL(b.name, object, newopts)
 }
 
+// TODO: Add a similar wrapper for GenerateSignedPostPolicyV4 allowing users to
+// omit PrivateKey/SignBytes
+
 func (b *BucketHandle) detectDefaultGoogleAccessID() (string, error) {
+	returnErr := errors.New("no credentials found on client and not on GCE (Google Compute Engine)")
+
 	if len(b.c.creds.JSON) > 0 {
 		var sa struct {
 			ClientEmail string `json:"client_email"`
 		}
-		// Don't error out if we can't marshal, fallback to GCE check.
 		err := json.Unmarshal(b.c.creds.JSON, &sa)
 		if err == nil && sa.ClientEmail != "" {
 			return sa.ClientEmail, nil
+		} else {
+			returnErr = errors.New("storage: empty client email in credentials")
+			if err != nil {
```

Comment:
Can we rearrange this if/else to remove the nesting? I think there are 3 cases, this seems clearer to me:

```
if err == nil && sa.ClientEmail != "" {
  return ...
} else if err != nil {
  returnErr = err
} else {
  returnErr = errors.New...
}
```

Same thing on L302 and following below.